DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP proposes a radically simple approach by deciding that all functions are first class functions. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = function ReturnType(Parameters) {
+  // Function body.
+};
```

Comment:
`f` can become an overload set. That is:
```
static assert (__traits(getOverloads, f_module, "f").length == 2);
```

AFAIK only `alias` can capture overload sets. So if this proposal is to be accepted, enums would also need to gain this ability.